### PR TITLE
migrate to TypeScript

### PR DESCRIPTION
**Pull request by Laurendus**

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -0,0 +1,13 @@
+{
+  "compilerOptions": {
+    "module": "CommonJS",
+    "target": "ES6",
+    "sourceMap": true
+  },
+  "exclude": [
+    "node_modules"
```
Comment (acran) [inline tsconfig.json:8]: Can you explain why this is needed? :)

File: index.ts [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+import popsicle = require('popsicle');
+import ClientOAuth2 = require('client-oauth2');
+import path = require("path");
+
+import morgan = require("morgan");
+import websocket = require("websocket");
+import bodyParser = require("body-parser");
+import dotenv = require("dotenv");
+import session = require("express-session");
+import {Express} from "express";
```
Comment (acran) [inline index.ts:13]: `express` is imported twice here, see first line

File: index.ts [hunk 3]
```diff
@@ -0,0 +1,308 @@
+import express = require("express");
```
Comment (acran) [inline index.ts:1]: the `import = require()` syntax is meant more as a compatibility to CommonJS, [see docs](https://www.typescriptlang.org/docs/handbook/modules.html#export--and-import--require); why not use
```suggestion
import * as express from 'express';
```
?

File: index.ts [hunk 4]
```diff
@@ -10,12 +10,13 @@ import websocket = require("websocket");
...
 import dotenv = require("dotenv");
 import session = require("express-session");
+import {Express} from "express";
 
 dotenv.config();
 
 
-var app = express();
-var server = new http.Server(app);
+var app: Express = express();
```
Comment (acran) [inline index.ts:18]: With proper types for the `express` module the type of `app` can be easily inferred here by typescript and is also for the reader, so it could be omitted here.

File: index.ts [hunk 5]
```diff
@@ -53,9 +54,9 @@ function getPropertyByPath(obj, path) {
     }, obj);
 }
 
-const oauthEnabled = !!process.env.OAUTH_CLIENT_ID;
+const oauthEnabled: boolean = !!process.env.OAUTH_CLIENT_ID;
```
Comment (acran) [inline index.ts:57]: the type here is trivially inferred and should be omitted here

File: index.ts [hunk 6]
```diff
@@ -69,7 +70,7 @@ function checkAuth(req, res, next) {
 }
 
 if (!oauthEnabled) {
-    app.get("/", function (req, res) {
+    app.get("/", function (req: express.Request, res: express.Response) {
```
Comment (acran) [inline index.ts:73]: for better readability:
~~~typescript
import {Request, Response} from 'express';
// ...
app.get("/", function (req: Request, res: Response) { //...
~~~

File: index.ts [hunk 7]
```diff
@@ -1,283 +1,291 @@
...
-var session = require("express-session");
+
+
 app.use(session({
-  resave: false,
-  saveUninitialized: false,
-  secret: process.env.COOKIE_SECRET,
+    resave: false,
+    saveUninitialized: false,
+    secret: process.env.COOKIE_SECRET,
```
Comment (acran) [inline index.ts:29]: This is one _massive_ commit but most of the diff comes purely from whitespace changes.
This makes reading the actual changes really hard, so please don't introduce unrelated whitespace changes - if you intend to do them, better do them in a separate style-only commit.

**Review by acran — CHANGES_REQUESTED** [review]
Renaming `index.js` breaks the `npm start` see [`package.json`](https://github.com/Laurendus/meeting-moderation-cards/blob/097f98500e2e8bd5f6a49e1b8f2f4e018ab1afba/package.json#L12). You'll need to first compile the `index.ts` e.g. with `tsc` or run it with e.g. [`ts-node`](https://www.npmjs.com/package/ts-node).

422c765ea093659e1ae318ba32968591b8f4126e contains mainly only whitespace changes which makes it hard to review the actual changes. Remove the whitespace changes or move them to a dedicated commit please.

File: index.ts (hunk 3 above)
Comment (Laurendus) [reply index.ts:1]: I'm afraid that I feel like I'm out of my depth here. I understand the linked section like my version ist the correct way. I explicitly changed to this syntax because of your section because I could not find any strong argument for other syntax variants.

File: index.ts (hunk 7 above)
Comment (Laurendus) [reply index.ts:29]: Actually my intention was to introduce a consistent styling which included 4 space indentation and importing before writing any code. This commit was intended to split this styling change from all the other commits and was therefore introduced first, to make the following changes more comprehensible. But for my IDE to detect and rearrange the imports correctly, I hat to change the syntax to `import abc = require('xyz');`.
This should have been the only real code change in this commit and I associated this to the topic 'coding style'.

File: tsconfig.json (hunk 1 above)
Comment (Laurendus) [reply tsconfig.json:8]: I do not know why source maps are needed other than the general statement that they make debugging easier so I left them as part of the default template.
The same goes for the exclusion of `node_modules` because I assumed that it is generally not a got idea to let typescript (re)compile included files. I am aware of the default exclusion when not explicitly specified but defaults can change and this is the only needed exclusion atm.

File: index.ts (hunk 2 above)
Comment (Laurendus) [reply index.ts:13]: The second import was necessary for the type annotation. Since I still haven't completely understood the import system, I am happy for clarifications.

**Comment by Laurendus** [comment]
> Renaming `index.js` breaks the `npm start` see [`package.json`](https://github.com/Laurendus/meeting-moderation-cards/blob/097f98500e2e8bd5f6a49e1b8f2f4e018ab1afba/package.json#L12). You'll need to first compile the `index.ts` e.g. with `tsc` or run it with e.g. [`ts-node`](https://www.npmjs.com/package/ts-node).

I disliked pushing a compiled result and assumed it would be standard practice to compile typescript before running the code hence no changes there. the [`ts-node`](https://www.npmjs.com/package/ts-node)-recommendation seems like a good idea.

**Comment by acran** [comment]
> I disliked pushing a compiled result and assumed it would be standard practice to compile typescript before running the code hence no changes there. the [`ts-node`](https://www.npmjs.com/package/ts-node)-recommendation seems like a good idea.
Committing the build product was not implied here ;)
But to run it with `node` it needs to be complied first, either explicitly (`tsc`) or implicitly (`ts-node`). Without that it currently does not run.

File: index.ts (hunk 3 above)
Comment (acran) [reply index.ts:1]: The latter syntax is the [recommended one](https://www.typescriptlang.org/docs/handbook/modules.html#import) since it is the native syntax for JavaScript modules introduced with ECMAScript2015.

File: index.ts (hunk 7 above)
Comment (acran) [reply index.ts:29]: > Actually my intention was to introduce a consistent styling which included 4 space indentation
But this is not clear from the commit and the diff shows massive changes - reviewing this is hard.

Comment (Laurendus) [reply index.ts:29]: I guess we can resolve this point by agreeing that style changes have nothing to do with the migration to typescript and therefore should not be part of this PR. I will see how I can elegantly undo these changes.
I would still keep the rearrangement of imports but with your [suggested style](https://github.com/BDSU/meeting-moderation-cards/pull/6#discussion_r541778653) (if possible)

File: index.ts [hunk 8]
```diff
@@ -31,12 +34,12 @@ app.engine(
...
     extname: "hbs",
-    defaultLayout: false,
+    defaultLayout: "",
   })
 );
 
-function getPropertyByPath(obj, path) {
-  return path.split('.').reduce((data, key) => {
+function getPropertyByPath(obj: object, path: string) {
+  return path.split(".").reduce((data, key) => {
```
Comment (acran) [inline index.ts:42]: Why is changing quotes here a _necessary_ change?

File: index.ts [hunk 9]
```diff
@@ -163,15 +196,19 @@ wss.on("request", function (request) {
...
         do {
           id = generateId();
         } while (room.connections.filter((item) => item.id == id).length > 0);
 
-        room.connections.push({ name: data.name, connection, id });
+        room.connections.push({
+          name: data.name,
+          connection: connection,
+          id: id,
+        });
```
Comment (acran) [inline index.ts:211]: What is this change required for? The line only gets longer and is the reformatted. Has it to do with the `typescript` switch?

File: index.ts [hunk 10]
```diff
@@ -123,22 +138,40 @@ app.all("/stimmung/:room?", checkAuth, function (req, res) {
       : "",
     html_author: process.env.HTML_AUTHOR ? process.env.HTML_AUTHOR : "",
     name: req.session.name,
-    room: req.params.room || '',
+    room: req.params.room || "",
```
Comment (acran) [inline index.ts:141]: see above

File: index.ts [hunk 11]
```diff
@@ -123,22 +138,40 @@ app.all("/stimmung/:room?", checkAuth, function (req, res) {
...
   });
 });
 
 server.listen(process.env.PORT, () => {
   console.log(`Server listening on port ${process.env.PORT}`);
 });
 
-var WSS = require("websocket").server;
+import * as websocket from "websocket";
```
Comment (acran) [inline index.ts:149]: Moving all `import`s to the top as in the previous draft is actually a good idea.

**Review by acran — COMMENTED** [review]
There are a bunch of reformattings going on here. But I can not quite follow based on what these take place...

File: index.ts (hunk 8 above)
Comment (Laurendus) [reply index.ts:42]: Wasn't a necessary change on my side. Must have been husky.

File: index.ts (hunk 9 above)
Comment (Laurendus) [reply index.ts:211]: husky as well. I did no such expansions.

File: index.ts (hunk 10 above)
Comment (Laurendus) [reply index.ts:141]: husky

**Review by acran — APPROVED** [review]
LGTM

File: package.json [hunk 12]
```diff
@@ -9,16 +9,23 @@
     "url": "https://github.com/mikel-jason/stimmungskarten.git"
   },
   "scripts": {
-    "start": "node index.js"
+    "start": "tsc && node index.js"
```
Comment (acran) [inline package.json:12]: you might want to add `index.js` to `.gitignore`

**Review by sarcaustech — APPROVED** [review]
